Datatypes refinement demo (reference only)

oetyng opened the pull request:
A demo of the full data types refinement change, for reference when reviewing the more granular draft PRs:

- https://github.com/maidsafe/safe-nd/pull/141
- https://github.com/maidsafe/safe-nd/pull/142
- https://github.com/maidsafe/safe-nd/pull/143
- https://github.com/maidsafe/safe-nd/pull/144

**Not to be merged.**

**NB:** The draft is primarily intended to focus on discussion about the **concept changes** and **idiomatic rust**. Small things, like typos or smallish wording changes in docs, indentation and similar trivialities can be addressed in the real PR (if they are still present). 
Anything off that is already present in current code base (i.e. not part of this change) could perhaps better go into a proper issue - as to keep discussion focused around the actual changes.

madadam committed:
refactor/adata: simplify the implementation

Use policy-based design instead of traits and macros.

madadam committed:
refactor/mdata: simplify the implementation

Use policy-based design instead of traits and macros.

oetyng committed:
refactor/datatypes: init move from ad to sequence

oetyng committed:
refactor/sequence: fix sequence index naming

oetyng committed:
refactor/sequence: add misc naming improvements

oetyng committed:
refactor/datatypes: move shared code from sequence

oetyng committed:
refactor/datatypes: copy map.rs from sequence.rs

oetyng committed:
feat/map: impl full private/public ad/md keyval

oetyng committed:
tests/map: impl crud happy path tests

oetyng committed:
tests/map: impl basic crud unhappy path tests

oetyng committed:
refactor/sequence: remove key-value semantics

oetyng committed:
refactor/datatypes: clean up the semantics
Many names/docs are confusing the concepts, these changes clear it up.

oetyng committed:
feat/map: give private history, add hard_del/updat
There are uses cases for keeping versions also when data is not public.

oetyng committed:
refactor/datatypes: impl fine grained permissions

oetyng committed:
tests/datatypes: impl fine grained permissions

oetyng committed:
refactor: add new modules, extract tests
Motivation: The file sizes are growing very large with thousands of LOC.

oetyng committed:
refactor/tests: move permissions tests to its mod
Code base needs organisation, to avoid thousands of LOC in files.

oetyng committed:
refactor/blob: renamed immutable data + misc clean

oetyng committed:
refactor: naming cleanup

oetyng committed:
refactor: add sequence cmd
Still a bit wip.

oetyng committed:
refactor/request: cleanup misc docs

oetyng committed:
refactor: rename permissions mod to access_control

oetyng committed:
refactor/access_control: clear up concepts

oetyng committed:
refactor/request: fix misleading names

oetyng committed:
refactor/auth: rename access_control to auth

oetyng committed:
refactor/auth: additional changes to auth

oetyng committed:
refactor: organize request components

oetyng committed:
feat/map: add key, owner and auth history + range

oetyng committed:
fix/map: remove dead code

oetyng committed:
docs/auth: add missing docs

oetyng committed:
fix/api: add missing apis to request and response

oetyng committed:
refactor/datatypes: replace 'index' with 'version'

oetyng committed:
refactor: use 'InvalidOperation' when invalid call

oetyng committed:
fix/request: add map tx commit + handle sequen cmd

oetyng committed:
feat/blob: add helper functions for BlobData enum

oetyng committed:
feat/kind: add `from_flag` method to `Kind` types

oetyng committed:
chore/ci: filter commits for deployment
We can use the `--no-merges` option to remove the merge
related commits.

oetyng committed:
fix/all: fix deprecation errors on rust 1.37

oetyng committed:
feat/permissions: revert back to coarse grained

oetyng committed:
feat/data: handle requests wip
Single entry point for requests,
piping of them through access control and further handling.

oetyng committed:
fix/request: refactor err response and fmt debug

oetyng committed:
feat/data: handle requests wip
Hard erasure, owner and permissions writes handling added.
Validation of owner for deletion added.

oetyng committed:
refactor/auth: rename to access_control/_list
Auth is ambiguous and not very descriptive.

oetyng committed:
chore: remove temporary allow of dead code

oetyng committed:
chore: add missing references

oetyng committed:
refactor: simplify request and access control
The previous attempt at read / write separation is deferred for now.

oetyng committed:
refactor/sequence: simplify append api

oetyng committed:
fix/sequence: add address to AppendOperation

oetyng committed:
fix/sequence: fix instantiation and pub address

oetyng committed:
refactor: granularize AppPermissions struct
- add perform_mutations and get_balance booleans

oetyng committed:
feat/request: add missing value
GetMapKeyHistories variant is missing in Request enum

oetyng committed:
fix/request: remove unused parameter

oetyng committed:
fix/request: remove unused parameter 2

oetyng committed:
fix/response: expect vec<u8> on map get

oetyng committed:
fix/sequence: add missing apis
Also structures map and sequence a bit and removes some redundant code.

oetyng committed:
fix/request: update missed lines

oetyng committed:
fix: clone owners and access list histories
Caveat: Probably non rust idiomatic implementations. Needs revision.

oetyng committed:
fix: add missing clone

oetyng committed:
feat/map: add get_at api

oetyng committed:
refactor/map: make api more expressive

oetyng committed:
feat/map: add get_values api

oetyng committed:
feat/map: add data_entries api

oetyng committed:
feat/map: add get_keys api

oetyng committed:
chore: fix clippy warnings from rust 1.40 - step 1

oetyng committed:
chore: fix clippy warnings from rust 1.40 - step 2

oetyng committed:
chore: fix clippy warnings from rust 1.40 - step 3

oetyng committed:
chore: fix clippy warnings from rust 1.40 - step 4

maqi commented on src/authorization/access_control.rs:
```diff
@@ -0,0 +1,209 @@
+}
+
+impl PublicAccessList {
+    fn is_allowed_(&self, user: &User, access: AccessType) -> Option<bool> {
```
not sure, however this `is_allowed_` nested whitin `is_allowed` doesn't seem a good practice?

oetyng replied:
Mm I agree it's an iffy naming practice. Could need some change, I don't remember quite now why it was done like this. Will look it up.

oetyng replied:
The problem is the lack of method overloading in rust.
The one with underscore is a private one that is called for a more specific case.
I tried putting all validation in the trait impl, but that looked very messy.
I think that underscore is the lesser bad of the two.
But some distinguishing word could be added of course.

jeanphilippeD replied:
Yes, adding a word expressing the more specific case would be good. I don't seem to have seen such naming otherwise in rust.

oetyng replied:
I was actually wrong, the private one is the less specific.
The hard part though is that there is no difference in semantics so the extra word will be very artificial.

I don't know how you usually do, but adding the types of the parameters in the function name is not a very elegant style. The extension of that is to not name things for what they represent, i.e. intent, (user for PublicKey) but always for what they are implemented with (implementation detail).

The public one checks if the user (represented as public key) access combo is allowed.
The private one checks if the user (represented by enum User) access combo is allowed.

Feel free to drop suggestions here.

jeanphilippeD replied:
If there is really no need for having different name (not 100% convinced, is_allowed is so very generic seem we could be a bit more specific, but also the function seem really simple so maybe it is fine (Yes, adding type parameter is not a great idea in general)).

I think you need not use `_` though:
https://doc.rust-lang.org/book/ch19-03-advanced-traits.html

> Fully Qualified Syntax for Disambiguation: Calling Methods with the Same Name
Nothing in Rust prevents a trait from having a method with the same name as another trait’s method, nor does Rust prevent you from implementing both traits on one type. It’s also possible to implement a method directly on the type with the same name as methods from traits.

oetyng replied:
Yep, I saw this as well. Just seemed like a lot of bloat with a couple of traits + impl, when an underscore solves it. But, it is more correct of course. 
Not sure here, because I'm not a fan of that underscore, but sometimes I think terse and slim wins over "correct".

maqi commented on src/data/blob.rs:
```diff
@@ -0,0 +1,337 @@
+    value: Vec<u8>,
+    /// Contains a set of owners of this data. DataHandlers enforce that a DELETE or OWNED-GET type
+    /// of request is coming from the ClientHandler of the owners.
+    owner: PublicKey,
```
shall use a vector?

oetyng replied:
You mean for PublicKey? I don't know, nothing changed there from previous implementation.

maqi replied:
Yes, for PublicKey.
well, the comment says `Contains a set of owners of this data`
maybe this shall be BlsKey ?

oetyng replied:
I think the end goal was this, to have it as `BlsKey`. I did ask around in slack a few weeks back for this, unfortunately I couldn't get it cleared out at the time. 
But I see no apparent reason why not doing it, I have always assumed `BlsKeys` will be the way we have multiple ownership of any and everything in the network.

oetyng replied:
Saw now that the question of PublicKey is moot, since it already is defined like this:

```
pub enum PublicKey {
    Ed25519(ed25519_dalek::PublicKey),
    Bls(threshold_crypto::PublicKey),
    BlsShare(threshold_crypto::PublicKeyShare),
}
```

But docs could be improved as you say.

maqi replied:
Ah, that will be great.

maqi commented on src/data/blob.rs:
```diff
@@ -0,0 +1,337 @@
+
+/// An immutable chunk of data which cannot be deleted.
+#[derive(Hash, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct PublicBlob {
```
maybe better have `PrivateBolb` & `PublicBolb` as:
```rust
struct Bolb {
    address: Address,
    value: Vec<u8>,
}

struct PrivateBolb {
    bolb: Bolb,
    owners: Vec<PublicKey>,
}

struct PublicBolb {
    bolb: Bolb,
}
```

oetyng replied:
Mm, perhaps, but it would break the pattern used for `Map` and `Sequence` (in code currently called `MapData` and `SequenceData`, a remnant of a previous a naming conflict, but "**Data**" is to be scrapped since the naming conflict was solved).

```
/// Object storing a Map variant.
pub enum MapData {
    /// Public instance with concurrency control.
    PublicSentried(PublicSentriedMap),
    /// Public instance.
    Public(PublicMap),
    /// Private instance with concurrency control.
    PrivateSentried(PrivateSentriedMap),
    /// Private instance.
    Private(PrivateMap),
}
```

I'd prefer keeping the pattern between the data types for expressing the same kind of phenomenon.

maqi commented on src/shared_data.rs:
```diff
@@ -0,0 +1,237 @@
+pub type Values = Vec<Value>;
+pub type Keys = Vec<Key>;
+
+/// Marker for sentried data.
```
any explanation of what is `sentried` and `non-sentried` data?

oetyng replied:
It's all described in the DataTypes Refinement Pre-RFC, there's both the HackMD and the PR to pre-rfc repo, I can link if you want?

maqi commented on src/shared_data.rs:
```diff
@@ -0,0 +1,237 @@
+
+// Set of data, owners, permissions versions.
+#[derive(Copy, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub struct ExpectedVersions {
```
why called `ExpectedVersions` ? this seems not much generic ?

oetyng replied:
It's per this design change: https://github.com/maidsafe/safe-nd/pull/126 (contains rationale)
... where recently (Sunday December 8th, in #safe-core) I discussed with @lionel1704 to change that to `Versions` instead of `Indices`:

> @lionel Regarding, the expected indices PR:
> 
> After having worked with the data types quite intensely now for a while, I've learned more about the usage of version and index at various places.
> I can see that it is used interchangeably for the same thing, where version is the concept which we IMO directly associate with the functionality of concurrency control and data perpetuity, and index is the implementation detail which reveals that we're working against a vector of sorts.
> I am quite strongly suggesting that we go consistent here and use only one of these all over, and also that it be version that we use (which was the case previously with MutableData), and keep the implementation detail from leaking through the layers.
> 
> This would in practice mean that the Index data struct is simply renamed to Version. And henceforth we talk about current version and expected version.
> I am absolutely positive that this corresponds better to the domain which it is used in (concurrency control & data perpetuity).
> Internally in safe-nd we can still map it to an Index type if we want to have a gradual transition to a language which best corresponds to that level of abstraction - i.e. when we're in the implementation details of the data types, we do know that we're dealing with indices.

where he replied

> You're absolutely right Edward. Index and version are used interchangeably. They can be renamed to version alone. Makes a lot of sense.

----------------------------------

When you say not much generic, could you please elaborate on what change you'd like to see and why?

maqi replied:
Ah, when I say `not much generic`, I just mean the name is `Expected`Version, meanwhile the content it holds is actually just version info, and could be used for not only `expected` purpose. (could be current, or last, ...)

oetyng replied:
Ah OK I see. Mm that's true. I think before it was often even just an `u64`, so at least it's a bit more readable now. But I guess static type checking on the actual purpose could be implemented as well.
Do you propose an `ExpectedVersion` type, that can only be used for that purpose?

maqi replied:
> Do you propose an ExpectedVersion type, that can only be used for that purpose?

Yes, I'd prefer to do so.

maqi commented on src/shared_data.rs:
```diff
@@ -0,0 +1,237 @@
+pub struct Owner {
+    pub public_key: PublicKey,
+    /// The expected Version of the data at the time this ownership change is to become valid.
+    pub expected_data_version: u64,
```
this means the data is owned by an owner only if the version of the data is greater than the expected_data_version? or less than it?
wondering what if a data's previous owner got removed and then added back later on?

just not clear about the usage of this `expected_data_version` and the below `expected_access_list_version`

oetyng replied:
If I understand your question correctly: no.

Except for naming, the implementation is the same now as before. I have also thought slightly about if something should change here, but it ended up fairly low on my priority list.

The Owner struct is used both as a DTO and as a domain model, which is a bit iffy IMO, because it is used for slightly different reasons, which are not apparent (as evidenced by your question as well).

As a DTO in a request to set the owner, the `expected_data_version` and `expected_access_list_version` acts as optimistic concurrency control, i.e. if anything changed in the data or permissions from the point that the requester sees it, to when the request is handled, it is rejected.
As a domain model, it tells the history of who the owner was at a particular data version. This would presumably let us answer questions like: `when current data version was x, who was the owner? when this set of users had this set of permissions (i.e. access list version y), who was the owner?`

maqi commented on src/shared_data.rs:
```diff
@@ -0,0 +1,237 @@
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum Address {
```
why we need to differentiate them?

oetyng replied:
I'm not entirely sure of this myself. It's also something kept from previous implementation.
I think the end result is that these get different name spaces, so you could store data under same xorname+tag for all four kinds. I'm not totally clear what benefits and drawbacks we have with that.
 Perhaps motivation is found in the RFCs for the Published/Unpublished, Sequenced/Unsequenced data types, otherwise, I think @nbaksalyar and others might have the exact background.

maqi commented on src/response.rs:
```diff
@@ -8,51 +8,74 @@
+    GetMapKeyHistoryRange(Result<MapValues>),
+    GetMapOwner(Result<Owner>),
+    GetMapOwnerAt(Result<Owner>),
+    GetMapOwnerHistory(Result<Vec<Owner>>),
```
this method sounds like trying to fetch the history of the owner, i.e. includes the owner that got removed.
or, it means to return the `owners list` ?

oetyng replied:
Ah yes, I had planned to fix this, `owners` field in `MapBase` should be called `owner_history`, and `access_list` in the same way `access_list_history`.

These are append only, so the act of removing an owner is to set a new owner.

Naming the list `owners` is a remnant of previous implementation, which gave the impression that they were all currently owners. But validation of ownership only ever looked at the last entry in the vector, so it is indeed a history of owners.

maqi commented on src/data/sequence.rs:
```diff
@@ -0,0 +1,788 @@
+    // This is the history of owners, with each entry representing an owner.  Each single owner
+    // could represent an individual user, or a group of users, depending on the `PublicKey` type.
+    owners: Vec<Owner>,
+    _flavour: S,
```
not used? then why we need this extra type parameter?

oetyng replied:
Good question. I would think it is to be able to differentiate the trait, and do different implementations based on the flavour.
Not sure why it would be stored in a field though, perhaps it has to for compiler to agree? I reused this solution from @madadam 's PR: https://github.com/maidsafe/safe-nd/pull/87

maqi commented on src/data/sequence.rs:
```diff
@@ -0,0 +1,788 @@
+{
+    /// Returns true if the provided access type is allowed for the specific user (identified y their public key).
+    pub fn is_allowed(&self, user: PublicKey, access: AccessType) -> bool {
+        if let Some(owner) = self.owner_at(Version::FromEnd(1)) {
```
this manual version value seems suspicious. 
at least a comment explaining why we can do so?

oetyng replied:
Yeah I know and agree. (Additionally, it might actually be FromEnd(0) to get last entry of the vec..)
The aim is top get the `current entry`, which is the current owner, without specifying an index.
It would only work if there are no race conditions. Do you have suggestion for this?

maqi replied:
I don't have a better solution to the issue, however, I'd at least use a const and put some comment to explain why doing so.
so that, in case we do need to change from `FromEnd(1)` to `FromEnd(0)`, it will be more explicit.

maqi commented on src/data/sequence.rs:
```diff
@@ -0,0 +1,788 @@
+
+    /// Returns true if the user is the current owner.
+    pub fn is_owner(&self, user: PublicKey) -> bool {
+        match self.owner_at(Version::FromEnd(1)) {
```
so only one owner allowed?

oetyng replied:
BLS key gives multiple owner under 1 key. This is just copied from previous implementation, nothing changed, except from wording. It was referred to as "owners" before, and `append_owner`, which would give impression of adding to existing owners, but validation of ownership was always made against _current entry_. The vector of owner was in fact the _history_ of owners, but it was not expressed in the naming of operations (therefore I renamed `append_owner` to `set_owner`). So, it has always been one owner-key always. Only with BLS shares can it currently be multiple owners. This needs clarification of how to be handled end to end, or if we want to do something differently.

maqi replied:
but the type of user is `PublicKey`, which normally means `a public key for just one user`
If we do want to use BLS key (or plan to do so later on), maybe change the type to something like `OwnershipKey`, and currently define `type OwnershipKey = PublicKey`, and later on switch it to BLSKey ?

oetyng replied:
Yep, I think that's a good idea.

oetyng replied:
Saw now that the question of PublicKey is moot, since it already is defined like this:

```
pub enum PublicKey {
    Ed25519(ed25519_dalek::PublicKey),
    Bls(threshold_crypto::PublicKey),
    BlsShare(threshold_crypto::PublicKeyShare),
}
```

maqi replied:
I am wondering that whether the network can allow different type of `PublicKey` co-exists within the network. (i.e. some node using `Ed25519` as PublicKey, meanwhile others using `Bls`).
Also, not sure about the diff between `Bls` and `BlsShare`, can they co-exist?
So, maybe making `PublicKey` as enum is un-necessary?

bochaco replied:
I'd also like to get a clarification about this, or do we need to use specific type of keys in different areas of the network, e.g. a SafeKey (CoinBalance) always is on a BLS PK, so all code related to managing this type of data uses a PublicKey::Bls/BlsShare, etc.

Moreover, why is it we need the Ed25519? can't we use BLS for everything?

maqi commented on src/data/map.rs:
```diff
@@ -0,0 +1,1561 @@
+    }
+
+    fn hard_apply(&mut self, operations: SentriedOperations) -> Result<()> {
+        let (insert, update, delete) = operations;
```
this usage is bit risky, especially `insert` and `update` has the same type.
can we make `SentriedOperations` as a proper struct and having a function returns insert/update/delete ?

oetyng replied:
Yeah, that would seem better, I'll try something out for it.

maqi commented on src/data/map.rs:
```diff
@@ -0,0 +1,1561 @@
+                        let _ = delete.insert(key.clone());
+                    }
+                };
+                (insert, update, delete)
```
will this lose the sequence of interleaving?
for example `insert A, delete A, insert A, update A` will be applied as: insert A, then update A, then delete A ?

oetyng replied:
Yes well spotted, this is a problem discussed in the pre-rfc here: https://github.com/maidsafe/pre-rfc/pull/112#discussion_r351410563
I suggest there that it should be fixed.

oetyng replied:
Yes well spotted, this is a problem discussed in the pre-rfc.
I suggest there that it should be fixed:

> We would need to consider what behaviour we want to support, and adapt network side handling to it. If we want to be able to do this:
> 
> ```
> let mymap = new Map();
> mymap.startTx();
> mymap.insert("key1", "value1");
> mymap.delete("key1");
> mymap.insert("key1", "value2"); // <---- insert same key again
> mymap.commit();
> ```
> 
> or even
> 
> ```
> let mymap = new Map();
> mymap.startTx();
> mymap.insert("key1", "value1");
> mymap.delete("key1");
> mymap.insert("key1", "value1"); // <--  insert same key again, with same value.
> mymap.commit();
> ```
> 
> .. then we have to change some of the current logics.
> Since it applies all inserts in a group, then all updates and then all deletes, the first example would blow up since it tries to perform
> ```
> mymap.insert("key1", "value1");
> mymap.insert("key1", "value2");
> mymap.delete("key1");
> ```
> 
> But, that's not a big thing to fix, we would then group operations by key, and _then_ apply them in the order they were added to the tx.
> 
> The second example would not result in "key1" = "value1", even if we fixed the key-grouping as per above, since the last operation - with current logics - would be filtered out, so we would end up with the key being deleted. Also not a big thing to fix, but we just need to be aware of things like this :)
>

oetyng committed:
refactor/datatypes: rename enums
As the naming conflict of the traits has been solved previously,
we can now remove "Data" from the type names.

oetyng committed:
docs/blob: fix minor errors and remnants

oetyng committed:
refactor/map: use struct instead of type alias
The parts of the Operation type alias could be easier to mixup.

oetyng committed:
chore/licence: update copyright year

oetyng committed:
refactor/version: use const for current version

jeanphilippeD commented on src/authorization/access_control.rs:
```diff
@@ -0,0 +1,209 @@
+
+impl AccessListTrait for PublicAccessList {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
+        match self.is_allowed_(&User::Specific(*user), access) {
```
Probably more idomatic and clearer:

```rust
     fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
        self.is_allowed_(&User::Specific(*user), access)
            .or_else(|| self.is_allowed_(&User::Anyone, access))
            .unwrap_or(false)
    } 
```

oetyng replied:
This is exactly the kind of comments I was hoping for. Great!

oetyng replied:
Aah, bad me..! I didn't check the logics or run the tests before committing this change :). 

It actually does not do the same thing.

oetyng committed:
refactor: idiomatic rust and clearer code

jeanphilippeD commented on src/authorization/access_control.rs:
```diff
@@ -0,0 +1,209 @@
+impl AccessListTrait for PrivateAccessList {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
+        match self.access_list.get(user) {
+            Some(access_status) => access_status.clone().is_allowed(access),
```
So like the other comment, a lot of time when reaching for a `match` it is better to look for chaining functions.
I'll let you look for the other cases. the option doc is pretty good here https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or

can think a bit of option as an iterator of one or zero element and treat them the same (can even convert option into an iterator if needed).

```rust
     fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
        // is the clone really needed?
        self.access_list.get(user)
            .map(|access_status| access_status.clone().is_allowed(access))
            .unwrap_or(false)
    }
```

jeanphilippeD commented on src/data/map.rs:
```diff
@@ -0,0 +1,1593 @@
+    pub fn is_owner(&self, user: PublicKey) -> bool {
+        use Map::*;
+        match self {
+            PublicSentried(data) => data.is_owner(user),
```
considering it is repeated again and again, might be worth using a macro like https://github.com/maidsafe/routing/blob/fleming/src/state_machine.rs#L36 (but a bit simpler), it will make all these function trivial and avoid duplication.

oetyng committed:
fix/version: use fromend(1) for current version

oetyng committed:
fix/chunk: rename blob to chunk
The previous implementation was mistaken for the Blob data type,
when it in fact was a Chunk.

oetyng committed:
test/response: uncomment and adjust tests

oetyng committed:
refactor/access: idiomatic rust and clearer code

oetyng committed:
fix: remove explicit closure for copying elements

oetyng committed:
refactor/kvpair: rename to KeyValuePair

oetyng committed:
refactor/map: use macro for repeated variant match

oetyng committed:
refactor/seq: use macro for repeated variant match

oetyng committed:
refactor/flavours: rename Sentried to Guarded

oetyng committed:
refactor/mods: organize tests more idiomatically

oetyng committed:
refactor/datatypes: check concurrency with param

oetyng commented:
This is (roughly) superseded by PR #163 (and expected PRs for coming CRDT data types).
A new take on the peripheral changes included in this PR needs to be done after CRDT data types are in place.
